server/src/main/java/org/apache/druid/server/coordinator/ReservoirSegmentSampler.java

```diff
@@ -19,36 +19,93 @@
 
 package org.apache.druid.server.coordinator;
 
+import com.google.common.math.IntMath;
+import org.apache.druid.client.ImmutableDruidServer;
 import org.apache.druid.timeline.DataSegment;
 
+import java.math.RoundingMode;
+import java.util.Collection;
 import java.util.List;
+import java.util.Spliterator;
 import java.util.concurrent.ThreadLocalRandom;
+import java.util.stream.Collectors;
+import java.util.stream.StreamSupport;
 
 final class ReservoirSegmentSampler
 {
+  private static final int SPLITERATOR_SIZE_THRESHOLD = 25;
 
   static BalancerSegmentHolder getRandomBalancerSegmentHolder(final List<ServerHolder> serverHolders)
   {
-    ServerHolder fromServerHolder = null;
+    ServerHolder fromServerHolder = reservoirSampleServer(serverHolders);
+    if (fromServerHolder == null) {
+      return null;
+    }
+
     DataSegment proposalSegment = null;
-    int numSoFar = 0;
-
-    for (ServerHolder server : serverHolders) {
-      for (DataSegment segment : server.getServer().getSegments()) {
-        int randNum = ThreadLocalRandom.current().nextInt(numSoFar + 1);
-        // w.p. 1 / (numSoFar+1), swap out the server and segment
-        if (randNum == numSoFar) {
-          fromServerHolder = server;
-          proposalSegment = segment;
+    Collection<DataSegment> segments = fromServerHolder.getServer().getSegments();
+
+    if (segments.size() == 1) {
+      proposalSegment = segments.iterator().next();
+    } else {
+      Spliterator<DataSegment> chosenSpliterator = segments.spliterator();
+      if (chosenSpliterator == null) {
+        return null;
+      }
+
+      for (int i = 0; i < IntMath.log2(segments.size(), RoundingMode.UP); i++) {
```

Comment:
can the logic here be flipped around like the following:

```java
while (chosenSpliterator.estimateSize() >= SPLITERATOR_SIZE_THRESHOLD) {
  Spliterator<DataSegment> newSpliterator = chosenSpliterator.trySplit();
  if (ThreadLocalRandom.current().nextBoolean()) {
    chosenSpliterator = newSpliterator;
  }
}
List<DataSegment> finalList = StreamSupport
              .stream(chosenSpliterator, false)
              .collect(Collectors.toList());
proposalSegment = getRandomElementFromList(finalList);
```

IMHO it is a bit easier to read

Reply:
This is right. The loop doesn't need to be bounded at all. Also, a single `randomBits = nextInt()` can be computed before the loop and checking `((randomBits >>= 1) & 1) == 0` instead of calling into TLR every time.

Reply:
Thank you both for the helpful suggestions! Have incorporated both of these comments.